File: VirtualPetC#/VirtualPetC#/Core/Pets/Bird.cs
namespace VirtualPetC_.Core.Pets;

/// <summary>
/// Bird class 
/// Special Ability: Song - singing boosts all stats slightly.
/// </summary>
public class Bird : BasePet
{
    private int songsPerformed;
    private double songCooldownRemaining;

    // Constructor - calls base Pet constructor
    public Bird(string name) : base(name)
    {
        songsPerformed = 0;
        songCooldownRemaining = 0;
    }

    public int SongsPerformed => songsPerformed;
    public double SongCooldownRemaining => songCooldownRemaining;

    
    
    public override void MakeSound()
    {
        Console.WriteLine($"{Name} says: Chirp chirp! 🐦");
        Happiness += 4;
        Console.WriteLine($"{Name} flutters wings happily!");
    }

    
    // Returns description of bird's special ability.
    
    public override string GetSpecialAbility()
    {
        string status = songCooldownRemaining > 0
            ? $"(Cooldown: {(int)songCooldownRemaining}s)"
            : "(Ready)";
        return $"Song - Boosts all stats! {status}";
    }

    public override string Play()
    {
        // Birds get balanced stats from flying - between dog and cat
        Happiness += 22;  // Between dog (25) and cat (18)
        Hunger -= 8;      // Between dog (10) and cat (7)

        return $"{Name} is flying around the room! \n{Name} does aerial acrobatics and loops in the air!\n{Name} lands gracefully, chirping with joy!";
    }

    
    // Special bird ability: Sing a beautiful song to boost all stats.
    public string UseSpecialAbility()
    {
        Hunger += SongHungerBoost;
        Happiness += SongHappinessBoost;
        Health += SongHealthBoost;
        Cleanliness += SongCleanlinessBoost;

        // Set cooldown
        songCooldownRemaining = SongCooldown;
        songsPerformed++;

        return $"{Name} sings a beautiful song! All stats boosted!";
    }

    
    // Check if the bird can use its special ability.
    public bool CanUseAbility()
    {
        return songCooldownRemaining <= 0;
    }

    
    // Update method to decrease song cooldown.
    public override void Update(double deltaTime)
    {
        base.Update(deltaTime);

        // Decrease cooldown
        if (songCooldownRemaining > 0)
        {
            songCooldownRemaining -= deltaTime;
            if (songCooldownRemaining < 0)
            {
                songCooldownRemaining = 0;
            }
        }
    }
}

File: VirtualPetGo/pet/bird.go
package pet

import "fmt"

type Bird struct {
	BasePet
	featherColor string
	songCooldown float64
}

func NewBird(name string, featherColor string) *Bird {
	return &Bird{
		BasePet:      newBasePet(name),
		featherColor: featherColor,
		songCooldown: 0,
	}
}

func (b *Bird) MakeSound() string {
	return "Chirp chirp"
}

func (b *Bird) Interact() string {
	return b.GetName() + "is chirping happily! " + b.MakeSound()
}

func (b *Bird) Play() string {
	b.BasePet.Play()
	// Reduced hunger decrease for bird
	b.setHunger(b.GetHunger() + 2)
	return b.GetName() + " performs aerial acrobatics! Happiness increased."
}

func (b *Bird) Update(deltaTime float64) {
	b.BasePet.Update(deltaTime)

	if b.songCooldown > 0 {
		b.songCooldown -= deltaTime
		if b.songCooldown < 0 {
			b.songCooldown = 0
		}
	}
}

func (b *Bird) UseSpecialAbility() string {
	b.setHunger(b.GetHunger() + SongHungerBoost)
	b.setHappiness(b.GetHappiness() + SongHungerBoost)
	b.setHealth(b.GetHealth() + SongHealthBoost)
	b.setCleanliness(b.GetCleanliness() + SongCleanlinessBoost)

	//Set Cooldown
	b.songCooldown = SongCooldown

	return b.GetName() + " sings a beautiful song! All boosted!"
}

func (b *Bird) CanUseAbility() bool {
	return b.songCooldown <= 0
}
func (b *Bird) GetStatus() Status {
	// Determine ability status text
	abilityStatus := "(Ready!)"
	if b.songCooldown > 0 {
		abilityStatus = fmt.Sprintf("(Cooldown: %.0f seconds)", b.songCooldown)
	}

	return Status{
		Name:     b.GetName(),
		Type:     "Bird",
		Age:      b.GetAge(),
		AgeStage: b.BasePet.getAgeStage(),

		// Core stats
		Health:      b.GetHealth(),
		Hunger:      b.GetHunger(),
		Happiness:   b.GetHappiness(),
		Cleanliness: b.GetCleanliness(),

		// Special ability info
		SpecialAbility: "Song - Boosts all stats!",
		AbilityStatus:  abilityStatus,

		// Overall status
		StatusMessage: b.BasePet.getStatusMessage(),
		IsAlive:       b.IsAlive(),
	}
}
